src/fancy-tools/brushTool.js

```diff
@@ -0,0 +1,295 @@
+
+    context.setTransform(1, 0, 0, 1, 0, 0);
+
+    const { cornerstone } = external;
+    const mouseCoordsCanvas = cornerstone.pixelToCanvas(element, this._lastImageCoords);
+    const canvasTopLeft = cornerstone.pixelToCanvas(element, {
+      x: 0,
+      y: 0
+    });
+    const radiusCanvas = cornerstone.pixelToCanvas(element, {
+      x: radius,
+      y: 0
+    });
+    const circleRadius = Math.abs(radiusCanvas.x - canvasTopLeft.x);
+
+    context.beginPath();
+    context.strokeStyle = color;
+    context.ellipse(mouseCoordsCanvas.x, mouseCoordsCanvas.y, circleRadius, circleRadius, 0, 0, 2 * Math.PI);
+    context.stroke();
+  }
+
+
+  /**
+   * Paints the data to the canvas.
+   *
+   * @private
+   * @param  {Object} eventData The data object associated with the event.
+   */
+  _paint (eventData) {
+    const configuration = this.configuration;
+    const element = eventData.element;
+    const { rows, columns } = eventData.image;
+    const { x, y } = eventData.currentPoints.image;
+    let toolData = getToolState(element, this._referencedToolData);
+
+    let pixelData;
+
+    if (toolData) {
+      pixelData = toolData.data[0].pixelData;
+    } else {
+      pixelData = new Uint8ClampedArray(eventData.image.width * eventData.image.height);
+      addToolState(element, this._referencedToolData, { pixelData });
+      toolData = getToolState(element, this._referencedToolData);
+    }
+
+    const brushPixelValue = configuration.draw;
+    const radius = configuration.radius;
+
+    if (x < 0 || x > columns ||
+      y < 0 || y > rows) {
+      return;
+    }
+
+    const pointerArray = getCircle(radius, rows, columns, x, y);
+
+    drawBrushPixels(pointerArray, pixelData, brushPixelValue, columns);
+
+    toolData.data[0].invalidated = true;
+
+    external.cornerstone.updateImage(eventData.element);
+  }
+
+  /**
+  * Event handler for NEW_IMAGE event.
+  *
+  * @event
+  * @param {Object} evt - The event.
+  */
+  _onNewImageCallback (evt) {
+    const eventData = evt.detail;
+    const element = eventData.element;
+    let toolData = getToolState(element, this._referencedToolData);
+
+    if (!toolData) {
+      const pixelData = new Uint8ClampedArray(eventData.image.width * eventData.image.height);
+
+      addToolState(element, this._referencedToolData, { pixelData });
+      toolData = getToolState(element, this._referencedToolData);
+    }
+
+    toolData.data[0].invalidated = true;
+    this._newImage = true;
+
+    external.cornerstone.updateImage(eventData.element);
+  }
+
+  /**
+   * Callback for when the tool is activated.
+   *
+   * @override
+   * @param {Object} element - The element to add/remove event listners.
+   */
+  activeCallback(element) {
+    const config = this.configuration;
+    this._changeDrawColor(config.draw);
+
+    element.removeEventListener(EVENTS.NEW_IMAGE, this._onNewImageCallback);
+    element.addEventListener(EVENTS.NEW_IMAGE, this._onNewImageCallback);
```

Comment:
We check for a `newImageCallback` on each tool now. Use that instead, and then check your tool's `mode` property to see if you should execute the logic in that hook?

Reply:
Aha, didn't realise a dispatcher for this method was in now. I'll fix this for the brushTool for now, but as I mentioned elsewhere, perhaps this rendering functionality needs to be centralized somewhere away from a specific brush tool.

Reply:
EDIT: I need to use newImageCallback even when the tool is passive/enabled. This is because the tool uses this hook in order to redraw the element to use the new slices toolData. I'm not sure if there is an easier way around this now, this is how we had to do it in the v2 implementation.

Reply:
I won't change anything about this comment until we have discussed this.

Reply:
@JamesAPetts I'm okay with updating the `newImageEventDispatcher` to include tools with `passive`, and `enabled` modes (like the `render` dispatcher). Then we can allow each tool that implements the callback decide which modes it will care about that event for.

Unlike the other callbacks, this one isn't traditionally prevented/stopped from reaching other tools, so that should be a safe and reasonable approach.

Reply:
Alrighty, I'll make this update tomorrow then.